storage: fix Allocator.{RebalanceTarget,ShouldRebalance}

petermattis opened the pull request:
Allocator.ShouldRebalance previously decided if the lease-holder replica
store was a candidate for rebalancing, but this didn't play well with the
recently added heuristic to not remove the lease-holder replica due to the
small blip of unavailability it causes. ShouldRebalance has now been
integrated with RebalanceTarget which checks to see if any of the replicas
other than the lease-holder are candidates for rebalancing (range count
greater than the mean).

Fixes #8112

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8124)

<!-- Reviewable:end -->

petermattis committed:
storage: fix Allocator.{RebalanceTarget,ShouldRebalance}

Allocator.ShouldRebalance previously decided if the lease-holder replica
store was a candidate for rebalancing, but this didn't play well with
the recently added heuristic to not remove the lease-holder replica due
to the small blip of unavailability it causes. ShouldRebalance has now
been integrated with RebalanceTarget which checks to see if any of the
replicas other than the lease-holder are candidates for
rebalancing (range count greater than the mean).

Fixes #8112

petermattis committed:
storage: return the current range count in reservation responses

petermattis committed:
storage: remove unused filter parameter from AllocateTarget

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1, 4 of 4 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/allocator.go, line 282 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrfKB9RHlkOjBsUiZw:-KNrfKB9RHlkOjBsUiZx:275504825) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/allocator.go#L282)):*
> ```Go
> 	required roachpb.Attributes,
> 	existing []roachpb.ReplicaDescriptor,
> 	leaseStoreID roachpb.StoreID,
> ```

any reason for moving the position of this argument?

---

*[storage/balancer.go, line 32 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrg46lwbtQ54q144jm:-KNrg46lwbtQ54q144jn:1084868796) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/balancer.go#L32)):*
> ```Go
> type nodeIDSet map[roachpb.NodeID]struct{}
> 
> func formatCandidates(
> ```

can you add a test for this? i have no idea what the output looks like.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNrglTnxw5o8jRgyGAS:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/allocator.go, line 282 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrfKB9RHlkOjBsUiZw:-KNrjkFVg109mJqz3A2e:630600793) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/allocator.go#L282)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> any reason for moving the position of this argument?
</details>
Consistency with the argument ordering of `RemoveTarget`.  I need to update the comment for this method. It is no longer accurate.

---

*[storage/balancer.go, line 32 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrg46lwbtQ54q144jm:-KNrjzTF1b_1jLs5NblP:-872319888) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/balancer.go#L32)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you add a test for this? i have no idea what the output looks like.
</details>
`[<StoreID>:<RangeCount>[*]...]`

The `selected` candidate has `*` suffix. This is only used for logging. Do you think it rises to the level of needing a test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/allocator.go, line 282 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrfKB9RHlkOjBsUiZw:-KNro6sh_agZyTxT7nT8:1461765252) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/allocator.go#L282)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Consistency with the argument ordering of `RemoveTarget`.  I need to update the comment for this method. It is no longer accurate.
</details>
Indeed.

---

*[storage/balancer.go, line 32 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrg46lwbtQ54q144jm:-KNrnh4uSrEhgc-HRTbp:-857493668) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/balancer.go#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `[<StoreID>:<RangeCount>[*]...]`
> 
> The `selected` candidate has `*` suffix. This is only used for logging. Do you think it rises to the level of needing a test?
</details>
I guess I don't feel strongly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/allocator.go, line 282 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNrfKB9RHlkOjBsUiZw:-KNrstM3V2VkAGFIGRya:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb83e3e022ea276a808722445e8c083e6ded5a80/storage/allocator.go#L282)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Indeed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@BramGruneir, @cuongdo I'm going to wait for one of you to LGTM as well.

cuongdo commented:
First pass. I need more time to think through this change than I have this morning. I'm happy that we're now returning `range_count` as part of `ReservationResponse`!

Also, have you tested this change against a real or test cluster?

---

Reviewed 1 of 4 files at r1, 4 of 4 files at r6, 3 of 4 files at r7.
Review status: 8 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/balancer.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvhhxv03Oy1041y8eU:-KNvhhxw9njz56xqH0Mf:1520857582) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L78)):*
> ```Go
> ) *roachpb.StoreDescriptor {
> 	// Consider a random sample of stores from the store list.
> 	sl.stores = selectRandom(rcb.rand, 3, sl, excluded)
> ```

I know that the `3`is carried over from the existing code, but I think we should bump it up to `10` or something like that. I don't see any downside, and `3` is a pretty small number. `10` should suffice for anything but large clusters.

---

*[storage/balancer.go, line 114 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvl4i0FLxtOTRzbU6i:-KNvl4i0FLxtOTRzbU6j:-803634371) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L114)):*
> ```Go
> ) *roachpb.StoreDescriptor {
> 	// Attempt to select a better candidate from the supplied list.
> 	sl.stores = selectRandom(rcb.rand, 3, sl, excluded)
> ```

Maybe bump `3` to `10` or some larger number? I know that this is existing code, but while you're in here...

---

*[storage/balancer.go, line 147 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvlXncc6w_QEOxco97:-KNvlXncc6w_QEOxco98:2108800549) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L147)):*
> ```Go
> 	// Moving a replica from the given store makes its range count converge on
> 	// the mean range count.
> 	if store.Capacity.FractionUsed() <= maxFractionUsedThreshold &&
> ```

I don't think `store.Capacity.FractionUsed() <= maxFractionUsedThreshold` is needed for this check. It may actually be harmful. If `store`is greater than 95% full, it could benefit from the removal of some stores.

---

*[storage/replicate_queue.go, line 108 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNviUk0ka-9-RAr5X_I:-KNviUk0ka-9-RAr5X_J:-1452964621) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/replicate_queue.go#L108)):*
> ```Go
> 		return true, priority
> 	}
> 	// See if there is a rebalancing opportunity present.
> ```

Maybe add a note here stating that unlike other range-related operations (e.g. reads & writes from/to the range), rebalancing decisions can now be made by replica in a range?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNvjWEQRQ_wpfPGU4y1:970327509)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
If you haven't tested against a cluster, I recommend running the `Rebalance_3To5Small` test as documented in`acceptance/allocator_test.go`.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNvmaMAlKfQkn5E0yQw:-139225027)*
<!-- Sent from Reviewable.io -->

petermattis committed:
storage: bump allocatorRandomCount from 3 to 10

Add TODO about using `Capacity.FractionUsed()` when considering a store
for rebalancing.

petermattis commented:
Review status: 1 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/balancer.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvhhxv03Oy1041y8eU:-KNvyqZcYOlIu5AV_iAC:-532375293) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L78)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I know that the `3`is carried over from the existing code, but I think we should bump it up to `10` or something like that. I don't see any downside, and `3` is a pretty small number. `10` should suffice for anything but large clusters.
</details>
Added an `allocatorRandomCount` constant and set it to `10`. One effect this will have on smaller clusters is that we won't actually have a random selection, but will instead select the range with the largest/smallest range count which feels like it could lead to a thundering herd. 

---

*[storage/balancer.go, line 114 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvl4i0FLxtOTRzbU6i:-KNvz8uX06UVPsgpNUrn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L114)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Maybe bump `3` to `10` or some larger number? I know that this is existing code, but while you're in here...
</details>
Done.

---

*[storage/balancer.go, line 147 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvlXncc6w_QEOxco97:-KNvz9m_XlBlaMP5sKie:31305048) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L147)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I don't think `store.Capacity.FractionUsed() <= maxFractionUsedThreshold` is needed for this check. It may actually be harmful. If `store`is greater than 95% full, it could benefit from the removal of some stores.
</details>
Heh, I tried remove that condition and tests started failing. I've added a TODO and will file an issue. Not sure I understand your last sentence. If the store is greater than 95% full it will always be a candidate for rebalancing no matter what its range count is.

---

*[storage/replicate_queue.go, line 108 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNviUk0ka-9-RAr5X_I:-KNvzmTUvghoqHUnZ0N6:-996888363) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/replicate_queue.go#L108)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Maybe add a note here stating that unlike other range-related operations (e.g. reads & writes from/to the range), rebalancing decisions can now be made by replica in a range?
</details>
True, but we've configured the queue to only process ranges for which we hold the range lease. Allowing any replica of a range to make rebalancing decisions seems like a pretty significant change. I'm not sure I understand all of the interactions that could occur. What benefit do you see to allowing that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I've been testing this change with my `zerosum` tool which causes lots of rebalancing. I'll try the `Rebalance_3To5Small` test.

Ugh, I can't get `Rebalance_3To5Small` working:

```
~ make test PKG=./acceptance TESTS=Rebalance_3To5Small TESTTIMEOUT=48h TESTFLAGS='-v -remote -key-name google_compute_engine -cwd allocator_terraform -tf.keep-cluster=failed'
go test -v  -i ./acceptance
go test  -run "Rebalance_3To5Small" -timeout 48h ./acceptance -v -remote -key-name google_compute_engine -cwd allocator_terraform -tf.keep-cluster=failed
I160730 09:54:51.649738 rand.go:76  Random seed: 8848536223231935360
=== RUN   TestRebalance_3To5Small
I160730 09:54:51.650331 acceptance/util_test.go:213  logging to /var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/clustertest_163243185
I160730 09:54:51.650346 acceptance/allocator_test.go:139  creating cluster with 3 node(s)
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
The module root could not be found. There is nothing to output.
+ terraform apply -var=num_instances=3 -var=stores=--store=/mnt/data0 -var=example_block_writer_instances=0 -var=cockroach_disk_type=pd-standard -no-color -var=key_name=google_compute_engine -state=rebal-3to5s-bigbose.tfstate -var=prefix=rebal-3to5s-bigbose
provider.google.account_file
  Enter a value:
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Warnings:

  * provider.google: "account_file": [DEPRECATED] Use the credentials field instead

No errors found. Continuing with 1 warning(s).

Error creating plan: 3 error(s) occurred:

* Resource 'google_compute_instance.cockroach' does not have attribute 'network_interface.0.access_config.0.assigned_nat_ip' for variable 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
* Resource 'google_compute_instance.cockroach' does not have attribute 'network_interface.0.access_config.0.assigned_nat_ip' for variable 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
* Resource 'google_compute_instance.cockroach' does not have attribute 'network_interface.0.access_config.0.assigned_nat_ip' for variable 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
The module root could not be found. There is nothing to output.
+ terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
The module root could not be found. There is nothing to output.
logs copied to /var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/clustertest_163243185
--- FAIL: TestRebalance_3To5Small (0.40s)
	allocator_test.go:141: failed:
		stdout:
		stderr: exit status 1
	farmer.go:180: not destroying; run:
		(cd /Users/pmattis/Development/go/src/github.com/cockroachdb/cockroach/acceptance/allocator_terraform && terraform destroy -force -state rebal-3to5s-bigbose.tfstate)
FAIL
exit status 1
FAIL	github.com/cockroachdb/cockroach/acceptance	0.424s
make: *** [test] Error 1
```

I have both `GOOGLE_PROJECT` and `GOOGLE_CREDENTIALS` set correctly. I think. Assistance would be appreciated.

---

Review status: 1 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNw0G8vK0XFAMfhQu4J:-1019882659)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
To run the rebalance test copy and paste the command line in the comments
near the top of allocator_test.go. It should work exactly as shown.
On Sat, Jul 30, 2016 at 9:56 AM Peter Mattis <notifications@github.com>
wrote:

> I've been testing this change with my zerosum tool which causes lots of
> rebalancing. I'll try the Rebalance_3To5Small test.
>
> Ugh, I can't get Rebalance_3To5Small working:
>
> ~ make test PKG=./acceptance TESTS=Rebalance_3To5Small TESTTIMEOUT=48h TESTFLAGS='-v -remote -key-name google_compute_engine -cwd allocator_terraform -tf.keep-cluster=failed'
> go test -v  -i ./acceptance
> go test  -run "Rebalance_3To5Small" -timeout 48h ./acceptance -v -remote -key-name google_compute_engine -cwd allocator_terraform -tf.keep-cluster=failed
> I160730 09:54:51.649738 rand.go:76  Random seed: 8848536223231935360
> === RUN   TestRebalance_3To5Small
> I160730 09:54:51.650331 acceptance/util_test.go:213  logging to /var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/clustertest_163243185
> I160730 09:54:51.650346 acceptance/allocator_test.go:139  creating cluster with 3 node(s)
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
> The module root could not be found. There is nothing to output.
> + terraform apply -var=num_instances=3 -var=stores=--store=/mnt/data0 -var=example_block_writer_instances=0 -var=cockroach_disk_type=pd-standard -no-color -var=key_name=google_compute_engine -state=rebal-3to5s-bigbose.tfstate -var=prefix=rebal-3to5s-bigbose
> provider.google.account_file
>   Enter a value:
> There are warnings and/or errors related to your configuration. Please
> fix these before continuing.
>
> Warnings:
>
>   * provider.google: "account_file": [DEPRECATED] Use the credentials field instead
>
> No errors found. Continuing with 1 warning(s).
>
> Error creating plan: 3 error(s) occurred:
>
> * Resource 'google_compute_instance.cockroach' does not have attribute 'network_interface.0.access_config.0.assigned_nat_ip' for variable 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
> * Resource 'google_compute_instance.cockroach' does not have attribute 'network_interface.0.access_config.0.assigned_nat_ip' for variable 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
> * Resource 'google_compute_instance.cockroach' does not have attribute 'network_interface.0.access_config.0.assigned_nat_ip' for variable 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> The module root could not be found. There is nothing to output.
> + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color example_block_writer
> The module root could not be found. There is nothing to output.
> logs copied to /var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/clustertest_163243185
> --- FAIL: TestRebalance_3To5Small (0.40s)
>     allocator_test.go:141: failed:
>         stdout:
>         stderr: exit status 1
>     farmer.go:180: not destroying; run:
>         (cd /Users/pmattis/Development/go/src/github.com/cockroachdb/cockroach/acceptance/allocator_terraform && terraform destroy -force -state rebal-3to5s-bigbose.tfstate)
> FAIL
> exit status 1
> FAIL    github.com/cockroachdb/cockroach/acceptance 0.424s
> make: *** [test] Error 1
>
> I have both GOOGLE_PROJECT and GOOGLE_CREDENTIALS set correctly. I think.
> Assistance would be appreciated.
> ------------------------------
>
> Review status: 1 of 9 files reviewed at latest revision, 5 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNw0G8vK0XFAMfhQu4J:-1019882659>*
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8124#issuecomment-236366580>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABffptCoQF8ggofV5i2Rq3ziqile5Xizks5qa1gVgaJpZM4JYXNI>
> .
>

petermattis commented:
That's exactly what I'm running.

On Sat, Jul 30, 2016 at 10:01 AM, Cuong Do <notifications@github.com> wrote:

> To run the rebalance test copy and paste the command line in the comments
> near the top of allocator_test.go. It should work exactly as shown.
> On Sat, Jul 30, 2016 at 9:56 AM Peter Mattis <notifications@github.com>
> wrote:
>
> > I've been testing this change with my zerosum tool which causes lots of
> > rebalancing. I'll try the Rebalance_3To5Small test.
> >
> > Ugh, I can't get Rebalance_3To5Small working:
> >
> > ~ make test PKG=./acceptance TESTS=Rebalance_3To5Small TESTTIMEOUT=48h
> TESTFLAGS='-v -remote -key-name google_compute_engine -cwd
> allocator_terraform -tf.keep-cluster=failed'
> > go test -v -i ./acceptance
> > go test -run "Rebalance_3To5Small" -timeout 48h ./acceptance -v -remote
> -key-name google_compute_engine -cwd allocator_terraform
> -tf.keep-cluster=failed
> > I160730 09:54:51.649738 rand.go:76 Random seed: 8848536223231935360
> > === RUN TestRebalance_3To5Small
> > I160730 09:54:51.650331 acceptance/util_test.go:213 logging to
> /var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/clustertest_163243185
> > I160730 09:54:51.650346 acceptance/allocator_test.go:139 creating
> cluster with 3 node(s)
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color
> example_block_writer
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color
> example_block_writer
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color
> example_block_writer
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color
> example_block_writer
> > The module root could not be found. There is nothing to output.
> > + terraform apply -var=num_instances=3 -var=stores=--store=/mnt/data0
> -var=example_block_writer_instances=0 -var=cockroach_disk_type=pd-standard
> -no-color -var=key_name=google_compute_engine
> -state=rebal-3to5s-bigbose.tfstate -var=prefix=rebal-3to5s-bigbose
> > provider.google.account_file
> > Enter a value:
> > There are warnings and/or errors related to your configuration. Please
> > fix these before continuing.
> >
> > Warnings:
> >
> > * provider.google: "account_file": [DEPRECATED] Use the credentials
> field instead
> >
> > No errors found. Continuing with 1 warning(s).
> >
> > Error creating plan: 3 error(s) occurred:
> >
> > * Resource 'google_compute_instance.cockroach' does not have attribute
> 'network_interface.0.access_config.0.assigned_nat_ip' for variable
> 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
> > * Resource 'google_compute_instance.cockroach' does not have attribute
> 'network_interface.0.access_config.0.assigned_nat_ip' for variable
> 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
> > * Resource 'google_compute_instance.cockroach' does not have attribute
> 'network_interface.0.access_config.0.assigned_nat_ip' for variable
> 'google_compute_instance.cockroach.*.network_interface.0.access_config.0.assigned_nat_ip'
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color
> example_block_writer
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color instances
> > The module root could not be found. There is nothing to output.
> > + terraform output -state=rebal-3to5s-bigbose.tfstate -no-color
> example_block_writer
> > The module root could not be found. There is nothing to output.
> > logs copied to
> /var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/clustertest_163243185
> > --- FAIL: TestRebalance_3To5Small (0.40s)
> > allocator_test.go:141: failed:
> > stdout:
> > stderr: exit status 1
> > farmer.go:180: not destroying; run:
> > (cd /Users/pmattis/Development/go/src/
> github.com/cockroachdb/cockroach/acceptance/allocator_terraform &&
> terraform destroy -force -state rebal-3to5s-bigbose.tfstate)
> > FAIL
> > exit status 1
> > FAIL github.com/cockroachdb/cockroach/acceptance 0.424s
> > make: *** [test] Error 1
> >
> > I have both GOOGLE_PROJECT and GOOGLE_CREDENTIALS set correctly. I think.
> > Assistance would be appreciated.
> > ------------------------------
> >
> > Review status: 1 of 9 files reviewed at latest revision, 5 unresolved
> > discussions, some commit checks failed.
> > ------------------------------
> >
> > *Comments from Reviewable
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNw0G8vK0XFAMfhQu4J:-1019882659
> >*
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/8124#issuecomment-236366580
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABffptCoQF8ggofV5i2Rq3ziqile5Xizks5qa1gVgaJpZM4JYXNI
> >
> > .
>
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8124#issuecomment-236366833>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF6f98nl7sESEwhxoc9A-bcKzdWg3-nwks5qa1kkgaJpZM4JYXNI>
> .
>

petermattis commented:
So what environment variables should I have set before running `TestRebalance_3To5Small`. Currently I have `GOOGLE_PROJECT` and `GOOGLE_CREDENTIALS`, the latter of which is the contents (not the path) to credentials I created on the GCE web console. I've also tried setting `GOOGLE_ACCOUNT_FILE` but that didn't help.

cuongdo commented:
Yeah your environment is fine. The test code should verify that before
starting.

The errors (module root not found) point to some problem with the directory
specified by 'cwd'. It's strange though. Are there any local mods to your
'acceptance/allocator_terraform'? The cwd is relative to your 'acceptance'
directory.
On Sat, Jul 30, 2016 at 10:18 AM Peter Mattis <notifications@github.com>
wrote:

> So what environment variables should I have set before running
> TestRebalance_3To5Small. Currently I have GOOGLE_PROJECT and
> GOOGLE_CREDENTIALS, the latter of which is the contents (not the path) to
> credentials I created on the GCE web console. I've also tried setting
> GOOGLE_ACCOUNT_FILE but that didn't help.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8124#issuecomment-236367652>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABffpjUdEEI5u5cL4ZJYlY8Knocr0l5Sks5qa10cgaJpZM4JYXNI>
> .
>

BramGruneir commented:
Reviewed 1 of 4 files at r1, 4 of 4 files at r9, 2 of 4 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[roachpb/api.proto, line 913 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNw9_dlqwOwVOWeYmem:-KNw9_dlqwOwVOWeYmen:702757999) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/roachpb/api.proto#L913)):*
> ```protobuf
>   // The current number of ranges on the target store. This is
>   // returned even if the reservation isn't filled.
>   optional int32 range_count = 2;
> ```

Does this number include the total including the current reservations?  It probably should.

I've got a 1/2 done PR in which I add a new stats proto to return.  But it requires a good amount of refactoring of how we deal with stats at the store level.  So this should be a great stopgap.

---

*[storage/balancer.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvhhxv03Oy1041y8eU:-KNwDO7KjO9RLVyt4AEB:726204379) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L78)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Added an `allocatorRandomCount` constant and set it to `10`. One effect this will have on smaller clusters is that we won't actually have a random selection, but will instead select the range with the largest/smallest range count which feels like it could lead to a thundering herd. 
</details>
There's a paper that shows that really you only need 2.  But there's absolutely no harm in using more as this is not really a time sensitive operation.

---

*[storage/reservation.go, line 29 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFD1aqKics7-pj8-r:-KNwFD1aqKics7-pj8-s:2064678312) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L29)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/util/syncutil"
> 	"github.com/cockroachdb/cockroach/util/timeutil"
> 	"github.com/gogo/protobuf/proto"
> ```

Nit: this include should be up grouped with the other external libraries.

---

*[storage/reservation.go, line 117 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFUlwIV34_zPZOrhp:-KNwFUlwIV34_zPZOrhq:2129068456) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L117)):*
> ```Go
> 	resp := roachpb.ReservationResponse{
> 		Reserved:   false,
> 		RangeCount: proto.Int32(int32(b.metrics.replicaCount.Count())),
> ```

I think we should include reserved replicas in this count as well.  It's rare for us to not fill a reservation.  This will help prevent herding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I have no local mods to `acceptance/allocator_terraform`.

petermattis commented:
Review status: 2 of 10 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 913 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNw9_dlqwOwVOWeYmem:-KNwJ8JWjaA9F5lvAwH6:-1414972430) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/roachpb/api.proto#L913)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Does this number include the total including the current reservations?  It probably should.
> 
> I've got a 1/2 done PR in which I add a new stats proto to return.  But it requires a good amount of refactoring of how we deal with stats at the store level.  So this should be a great stopgap.
</details>
No, it is powered by the `replicaCount` metric. Changed.

---

*[storage/balancer.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvhhxv03Oy1041y8eU:-KNwJqJj0uPNfLRqJKmf:-142776495) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L78)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> There's a paper that shows that really you only need 2.  But there's absolutely no harm in using more as this is not really a time sensitive operation.
</details>
How is there no harm in using more? As my comment points out, using more actually removes randomness in small clusters.

---

*[storage/reservation.go, line 29 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFD1aqKics7-pj8-r:-KNwK4DhRbL1Sg0GAv_A:1641813679) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L29)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Nit: this include should be up grouped with the other external libraries.
</details>
Done. Does anyone even look at the imports list anymore outside of code reviews?

---

*[storage/reservation.go, line 117 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFUlwIV34_zPZOrhp:-KNwKsi7BtY3DamIfyw6:1122587979) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L117)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I think we should include reserved replicas in this count as well.  It's rare for us to not fill a reservation.  This will help prevent herding.
</details>
Done. Maybe. We need more extensive testing of rebalancing heuristics. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 4 files at r9, 4 of 8 files at r12, 2 of 4 files at r13, 2 of 2 files at r14.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 913 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNw9_dlqwOwVOWeYmem:-KNwRM-uVceCAGZTBr0N:235880872) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/roachpb/api.proto#L913)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> No, it is powered by the `replicaCount` metric. Changed.
</details>
Why is that nullable now?

---

*[storage/store_pool.go, line 488 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwRkUenELs1UzEpXr_:-KNwRkUenELs1UzEpXra:-1378326292) ([raw file](https://github.com/cockroachdb/cockroach/blob/b24591cb5aa47b7d338aca201b155f021e8dc55e/storage/store_pool.go#L488)):*
> ```Go
> 	}
> 
> 	if resp.RangeCount != nil {
> ```

is it ever nil?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 913 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNw9_dlqwOwVOWeYmem:-KNwS9EL0sHw0xnMwqvT:-536269624) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/roachpb/api.proto#L913)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Why is that nullable now?
</details>
Upgrades. Old servers won't send this field and using the default value of `0` would be bad.

---

*[storage/store_pool.go, line 488 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwRkUenELs1UzEpXr_:-KNwSI7oiF3rb9Rs07JT:-1656817398) ([raw file](https://github.com/cockroachdb/cockroach/blob/b24591cb5aa47b7d338aca201b155f021e8dc55e/storage/store_pool.go#L488)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it ever nil?
</details>
Yep, it's `nil` if we're talking to an old server.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 913 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNw9_dlqwOwVOWeYmem:-KNwT9xkqWDpFlM4GFGT:-891318147) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/roachpb/api.proto#L913)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Upgrades. Old servers won't send this field and using the default value of `0` would be bad.
</details>
Gah. Can you add a comment? We may want to "flush" this the next time we issue a stop-the-world update.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[roachpb/api.proto, line 913 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNw9_dlqwOwVOWeYmem:-KNwcxfNEUTlo9itC4c5:1400074649) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/roachpb/api.proto#L913)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Gah. Can you add a comment? We may want to "flush" this the next time we issue a stop-the-world update.
</details>
Added a comment. I'm doubtful this will be worthwhile to change in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r12, 2 of 4 files at r13, 4 of 4 files at r16, 2 of 4 files at r17, 2 of 2 files at r18.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/balancer.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvhhxv03Oy1041y8eU:-KNwhYNTC9kPlpduk63J:981727848) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L78)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> How is there no harm in using more? As my comment points out, using more actually removes randomness in small clusters.
</details>
I meant theoretically.  But is randomness the goal?  

---

*[storage/reservation.go, line 29 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFD1aqKics7-pj8-r:-KNwhwAmOpCjgN8MtEuc:764372925) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Done. Does anyone even look at the imports list anymore outside of code reviews?
</details>
Not really.  We should just create a make check for it.

---

*[storage/reservation.go, line 117 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFUlwIV34_zPZOrhp:-KNwpbDl2YUh-V6sDzQC:-1945975929) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L117)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Done. Maybe. We need more extensive testing of rebalancing heuristics. 
</details>
Indeed, that's what the simulator is for.  But it needs some work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-:-KNwqaN6G6KWMtNgfno6:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/balancer.go, line 78 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNvhhxv03Oy1041y8eU:-KNwrIXsOncpLrqCrcxR:-157581363) ([raw file](https://github.com/cockroachdb/cockroach/blob/5340ebffdc16f494fbf5ee6723fa681a415c24f4/storage/balancer.go#L78)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I meant theoretically.  But is randomness the goal?  
</details>
Without randomness, thundering herd is more likely. Consider a 10th (empty) node being added to a cluster. When we allocate a new replica for this range, we'll have 7 candidate stores (3 of the stores already contain replicas). Of those 7 stores, the new one will always be chosen. On the other hand, the other 6 candidates are likely already balanced so maybe this isn't a problem. Regardless, let's discuss this outside of this PR.

---

*[storage/reservation.go, line 29 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124#-KNwFD1aqKics7-pj8-r:-KNwsElYgoN76W8zx4Dz:-769468584) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b1d3a888c34bd101c0cf50cb98d743ee4d8b37b/storage/reservation.go#L29)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Not really.  We should just create a make check for it.
</details>
Or we can just train ourselves to not worry about this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8124)*
<!-- Sent from Reviewable.io -->
